src/hotspot/os_cpu/bsd_aarch64/vm_version_bsd_aarch64.cpp

```diff
@@ -53,20 +53,36 @@ static bool cpu_has(const char* optional) {
 void VM_Version::get_os_cpu_info() {
   size_t sysctllen;
 
-  // hw.optional.floatingpoint always returns 1, see
-  // https://github.com/apple/darwin-xnu/blob/master/bsd/kern/kern_mib.c#L416.
+  // hw.optional.floatingpoint always returns 1.
+  // See https://github.com/apple/darwin-xnu/blob/main/bsd/kern/kern_mib.c#L744
   // ID_AA64PFR0_EL1 describes AdvSIMD always equals to FP field.
+  // See https://developer.arm.com/documentation/ddi0595/2021-12/AArch64-Registers/ID-AA64PFR0-EL1--AArch64-Processor-Feature-Register-0
   assert(cpu_has("hw.optional.floatingpoint"), "should be");
   assert(cpu_has("hw.optional.neon"), "should be");
   _features = CPU_FP | CPU_ASIMD;
 
-  // All Apple-darwin Arm processors have AES and PMULL.
-  _features |= CPU_AES | CPU_PMULL;
+  // All Apple-darwin Arm processors have AES, PMULL, SHA1 and SHA2.
+  _features |= CPU_AES | CPU_PMULL | CPU_SHA1 | CPU_SHA2;
 
-  // Only few features are available via sysctl, see line 614
-  // https://opensource.apple.com/source/xnu/xnu-6153.141.1/bsd/kern/kern_mib.c.auto.html
-  if (cpu_has("hw.optional.armv8_crc32"))     _features |= CPU_CRC32;
-  if (cpu_has("hw.optional.armv8_1_atomics")) _features |= CPU_LSE;
+  // Only few features are available via sysctl.
+  // See https://github.com/apple/darwin-xnu/blob/main/bsd/kern/kern_mib.c#L855
+  // Regarding SHA512 and SHA3, there are two sysctl flags for them, which
+  // might be invented at different times. We check both here.
+  // See https://git.tartarus.org/?p=simon/putty.git;a=blob_plain;f=unix/utils/arm_arch_queries.c;hb=HEAD
+  if (cpu_has("hw.optional.armv8_crc32")) {
+    _features |= CPU_CRC32;
+  }
+  if (cpu_has("hw.optional.armv8_1_atomics")) {
```

Comment:
Why aren't we using `hw.optional.arm.FEAT_LSE` instead of `hw.optional.armv8_1_atomics` as per https://developer.apple.com/documentation/kernel/1387446-sysctlbyname/determining_instruction_set_characteristics ?

Reply:
This is Apple AArch64. Is there any point to testing for FEAT_LSE?

Reply:
Hi @gerard-ziemski, thanks for your useful link.

As denoted by the title of this PR, I originally added the support for crypto feature detection, i.e. SHA1, SHA2, SHA512 and SHA3. Hence, I didn't touch other CPU features except updating the code style (for CRC32 and your mentioned LSE).

Regarding your question, I think it's okay to check `hw.optional.arm.FEAT_LSE` inside this patch, as it's a small improvement as I see it. But I suggest
1) still checking the legacy name `hw.optional.armv8_1_atomics` as well (as what we did for SHA512 and SHA3 in this PR), for the sake of backward compatibility.
2) checking both `hw.optional.AdvSIMD` and `hw.optional.neon` (ASIMD feature has the similar issue as well).

Hi @theRealAph , 
As shown by Apple developer document (shared by @gerard-ziemski), as parameters of function `sysctlbyname`, `hw.optional.armv8_1_atomics` is the legacy name for `hw.optional.arm.FEAT_LSE`. Hence, we can check `FEAT_LSE` to determine whether atomic instructions are available or not.

I'm afraid I didn't fully get your question.
Do you mean, the usage of atomic instructions is quite different on Apple AArch64 compared to Linux AArch64 (perhaps, atomic instructions are not generated on Apple AArch64 platform?), hence there is no need to support CPU_LSE detection on Apple AArch64?

Reply:
Sure, never mind. It's just that Apple always supports LSE, but that doesn't really matter.

Reply:
Hi @gerard-ziemski , I improved the ASIMD and LSE features detection in the way I illustrated in my previous comment. See the latest revision. Is that good to you? Thanks.